src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,320 @@
+                context.RegisterSymbolAction(context => AnalyzeSymbol(context, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureType), s_symbols);
+            });
+        }
+
+        private static bool TypeSymbolHasAttribute(ITypeSymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType)
+        {
+            ImmutableArray<INamedTypeSymbol> interfaces = symbol.Interfaces;
+            foreach (INamedTypeSymbol? anInterface in interfaces)
+            {
+                var interfaceMembers = anInterface.GetMembers();
+                if (interfaceMembers.Length == 0)
+                {
+                    // Only tag empty interfaces to prevent breaking changes in the future
+                    requiresPreviewFeaturesSymbolsToUsageType.GetOrAdd(symbol, PreviewFeatureUsageType.EmptyInterface);
+                    return true;
+                }
+            }
+
+            INamedTypeSymbol? baseType = symbol.BaseType;
+            if (baseType != null)
+            {
+                return HasPreviewAttribute(baseType, requiresPreviewFeaturesSymbols);
+            }
+
+            return false;
+        }
+
+        private static bool ContainingTypeIsPreview(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            INamedTypeSymbol? containingType = symbol.ContainingType;
+            if (containingType != null)
+            {
+                return HasPreviewAttribute(containingType, requiresPreviewFeaturesSymbols);
+            }
+
+            return false;
+        }
+
+        private static bool SymbolIsStaticAndAbstract(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            bool isStaticAndAbstract = symbol.IsStatic && symbol.IsAbstract;
+
+            return isStaticAndAbstract && !ContainingTypeIsPreview(symbol, requiresPreviewFeaturesSymbols);
+        }
+
+        private static bool PropertyOrMethodSymbolHasAttribute(ISymbol propertyOrMethodSymbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType, INamedTypeSymbol? runtimeFeatureType)
+        {
+            if (SymbolIsStaticAndAbstract(propertyOrMethodSymbol, requiresPreviewFeaturesSymbols))
+            {
+                if (runtimeFeatureType != null && HasPreviewAttribute(runtimeFeatureType, requiresPreviewFeaturesSymbols))
+                {
+                    requiresPreviewFeaturesSymbolsToUsageType.GetOrAdd(propertyOrMethodSymbol, PreviewFeatureUsageType.StaticAbstract);
+                    return true;
+                }
+            }
+
+            if (propertyOrMethodSymbol.IsImplementationOfAnyImplicitInterfaceMember(out ISymbol baseInterfaceMember))
```

Comment:
This is a new extension method. The existing method doesn't expose `baseInterfaceMember`. Instead it only returned a bool.